lapcs.adoc

```diff
@@ -434,122 +434,131 @@ is only guaranteed to cover the `lp64d` and `lp64s` variant, that is, `GRLEN` is
 narrower than GRLEN bits are widened to GRLEN bits, with the upper bits undefined.
 
-* An integral argument is passed in a GAR if there is one available.
+** An integral argument is passed in a GAR if there is one available.
 Otherwise, it is passed on the stack. If the argument is narrower than the
 containing GAR, the <<int_ext_rules, general rules of integral extensions>>
 applies.
 
-*GRLEN < w~arg~ ≤ 2 × GRLEN*
+* GRLEN < w~arg~ ≤ 2 × GRLEN
 
-* The argument is passed in a pair of GARs with adjacent numbers, with the
+** The argument is passed in a pair of GARs with adjacent numbers, with the
 lower-ordered GRLEN bits in the low-numbered register. If only one GAR
 is available, the lower-ordered GRLEN bits are passed in this register
 and the most-significant GRLEN bits are passed on the stack. If no GAR is
 available, the whole argument is passed on the stack.
 
 ==== Structures
 
-The storage location of structures during argument passing is also determined
-by w~arg~ and GRLEN:
+Upon function calls and returns, a structure argument's storage location
+is mainly determined by its size and the number of floating-point
+and/or integer members it contains. A structure argument can be passed
+in up to two registers if available.
+
+The storage layout of a structure containing other structures or arrays
+is the same as its flattened counterpart, where the member structures are
+replaced by its individual members and member arrays of length n (n > 0)
+are broken down into n consecutive members of its element type.
+
+NOTE: Empty structures or unions are zero-sized in C while they have
+the size of 1 byte in {Cpp}.
+
+For example, `struct { struct { double d[1]; } a[2]; }` and
+`struct { double d0; double d1; }` should have the same storage layout when
+passed as function parameters.
 
-NOTE: Zero-length array members in structures should be ignored by both C and
-C++ compilers.
+*Structures without floating-point members*
 
-*w~arg~ > 2 × GRLEN*
+* w~arg~ > 2 × GRLEN
 
-* The structure is passed by reference, i.e. replaced in the argument list
-with its memory address. If there is an available GAR, the address is passed
-in the GAR, otherwise it is passed on the stack.
+** The argument is passed by reference, i.e. replaced in the argument list with
+its memory address. If there is an available GAR, the address is passed in the
+GAR, otherwise it is passed on the stack.
 
-*0 < w~arg~ ≤ GRLEN*
+* GRLEN < w~arg~ ≤ 2 × GRLEN
 
-* If the structure contains only one or two fp32 / fp64 members,
-and the same number of FARs are available, the members are passed in
-the FARs respectively, with the first member in the lower-numbered
-register.
+** The argument is passed in a pair of GARs with adjacent numbers, with the
+lower-ordered GRLEN bits in the low-numbered register. If only one GAR is
+available, the lower-ordered GRLEN bits are passed in this register and the
+most-significant GRLEN bits are passed on the stack. If no GAR is available, the
```

Comment:
```suggestion
higher-ordered GRLEN bits are passed on the stack. If no GAR is available, the
```

For some more symmetry because the previous clause referred to the "lower-ordered" bits?